### ISSUE TITLE
prxs install mod_digest.c fails

### ISSUE THREAD

**Issue by ardgeo** [proftpd/proftpd#931]
### What I Did

Fresh install of Rasbian Buster

Installed via `apt-get install proftpd-dev` (basic + dev) has DSO Support.
Copied from proftpd git hub page the mod_digest.c file (as this is newest and has openssl support for 1.1.1d I beleive?)

Tried to run 
```
prxs -c -i -d mod_digest.c
```
(tried as user and as sudo)

Each time it fails

Error on build:
```
prxs -c -i -d mod_digest.c
libtool --mode=compile gcc  -DLINUX -g2 -g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-Q9YVRO/proftpd-dfsg-1.3.6=. -fstack-protector-strong -Wformat -Werror=format-security -Wall -fno-omit-frame-pointer -Werror=implicit-function-declaration -DPR_SHARED_MODULE -I. -I/usr/include/proftpd -c mod_digest.c
libtool: compile:  gcc -DLINUX -g2 -g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-Q9YVRO/proftpd-dfsg-1.3.6=. -fstack-protector-strong -Wformat -Werror=format-security -Wall -fno-omit-frame-pointer -Werror=implicit-function-declaration -DPR_SHARED_MODULE -I. -I/usr/include/proftpd -c mod_digest.c  -fPIC -DPIC -o .libs/mod_digest.o
mod_digest.c: In function ‘EVP_crc32’:
mod_digest.c:400:35: warning: passing argument 1 of ‘EVP_MD_meth_set_input_blocksize’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
   EVP_MD_meth_set_input_blocksize(md, CRC32_BLOCK);
                                   ^~
In file included from mod_digest.c:76:
/usr/include/openssl/evp.h:80:45: note: expected ‘EVP_MD *’ {aka ‘struct evp_md_st *’} but argument is of type ‘const EVP_MD *’ {aka ‘const struct evp_md_st *’}
 int EVP_MD_meth_set_input_blocksize(EVP_MD *md, int blocksize);
                                     ~~~~~~~~^~
mod_digest.c:401:31: warning: passing argument 1 of ‘EVP_MD_meth_set_result_size’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
   EVP_MD_meth_set_result_size(md, CRC32_DIGEST_LENGTH);
                               ^~
In file included from mod_digest.c:76:
/usr/include/openssl/evp.h:81:41: note: expected ‘EVP_MD *’ {aka ‘struct evp_md_st *’} but argument is of type ‘const EVP_MD *’ {aka ‘const struct evp_md_st *’}
 int EVP_MD_meth_set_result_size(EVP_MD *md, int resultsize);
                                 ~~~~~~~~^~
mod_digest.c:402:32: warning: passing argument 1 of ‘EVP_MD_meth_set_app_datasize’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
   EVP_MD_meth_set_app_datasize(md, sizeof(EVP_MD *) + sizeof(CRC32_CTX));
                                ^~
In file included from mod_digest.c:76:
/usr/include/openssl/evp.h:82:42: note: expected ‘EVP_MD *’ {aka ‘struct evp_md_st *’} but argument is of type ‘const EVP_MD *’ {aka ‘const struct evp_md_st *’}
 int EVP_MD_meth_set_app_datasize(EVP_MD *md, int datasize);
                                  ~~~~~~~~^~
mod_digest.c:403:24: warning: passing argument 1 of ‘EVP_MD_meth_set_init’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
   EVP_MD_meth_set_init(md, crc32_init);
                        ^~
In file included from mod_digest.c:76:
/usr/include/openssl/evp.h:84:34: note: expected ‘EVP_MD *’ {aka ‘struct evp_md_st *’} but argument is of type ‘const EVP_MD *’ {aka ‘const struct evp_md_st *’}
 int EVP_MD_meth_set_init(EVP_MD *md, int (*init)(EVP_MD_CTX *ctx));
                          ~~~~~~~~^~
mod_digest.c:404:26: warning: passing argument 1 of ‘EVP_MD_meth_set_update’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
   EVP_MD_meth_set_update(md, crc32_update);
                          ^~
In file included from mod_digest.c:76:
/usr/include/openssl/evp.h:85:36: note: expected ‘EVP_MD *’ {aka ‘struct evp_md_st *’} but argument is of type ‘const EVP_MD *’ {aka ‘const struct evp_md_st *’}
 int EVP_MD_meth_set_update(EVP_MD *md, int (*update)(EVP_MD_CTX *ctx,
                            ~~~~~~~~^~
mod_digest.c:405:25: warning: passing argument 1 of ‘EVP_MD_meth_set_final’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
   EVP_MD_meth_set_final(md, crc32_final);
                         ^~
In file included from mod_digest.c:76:
/usr/include/openssl/evp.h:88:35: note: expected ‘EVP_MD *’ {aka ‘struct evp_md_st *’} but argument is of type ‘const EVP_MD *’ {aka ‘const struct evp_md_st *’}
 int EVP_MD_meth_set_final(EVP_MD *md, int (*final)(EVP_MD_CTX *ctx,
                           ~~~~~~~~^~
mod_digest.c:406:27: warning: passing argument 1 of ‘EVP_MD_meth_set_cleanup’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
   EVP_MD_meth_set_cleanup(md, crc32_free);
                           ^~
In file included from mod_digest.c:76:
/usr/include/openssl/evp.h:92:37: note: expected ‘EVP_MD *’ {aka ‘struct evp_md_st *’} but argument is of type ‘const EVP_MD *’ {aka ‘const struct evp_md_st *’}
 int EVP_MD_meth_set_cleanup(EVP_MD *md, int (*cleanup)(EVP_MD_CTX *ctx));
                             ~~~~~~~~^~
mod_digest.c:407:25: warning: passing argument 1 of ‘EVP_MD_meth_set_flags’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
   EVP_MD_meth_set_flags(md, 0);
                         ^~
In file included from mod_digest.c:76:
/usr/include/openssl/evp.h:83:35: note: expected ‘EVP_MD *’ {aka ‘struct evp_md_st *’} but argument is of type ‘const EVP_MD *’ {aka ‘const struct evp_md_st *’}
 int EVP_MD_meth_set_flags(EVP_MD *md, unsigned long flags);
                           ~~~~~~~~^~
mod_digest.c: In function ‘get_key_for_cache’:
mod_digest.c:1372:3: error: implicit declaration of function ‘pr_snprintf’; did you mean ‘BIO_snprintf’? [-Werror=implicit-function-declaration]
   pr_snprintf(mtime_str, sizeof(mtime_str)-1, "%llu",
   ^~~~~~~~~~~
   BIO_snprintf
cc1: some warnings being treated as errors
prxs: error executing command (1)
```

### What I Expected/Wanted

I wanted to install mod_digest on my proftpd instance. It is the key reason i want to use proftpd.
However I did not find a precomplied binary that included mod_digest but I did see that DSO support should allow for an easy addition... but it appears to be broken?

### ProFTPD Version and Configuration

Installed on Rasbian Buster via sudo `apt-get install proftpd-dev` 
Copied down the mod_digest from this site rather than https://github.com/Castaglia/proftpd-mod_digest as that is 4 years old and I know doesn't support openssl versions.
Reinstalled libssl-dev using apt-get to make sure i had a copy of evp.h in /usr/include/openssl/ as it wasn't originally there...

`echo $?` produced a 1 after the prxs
```
    $ proftpd -V

`proftpd -V
Compile-time Settings:
  Version: 1.3.6 (stable)
  Platform: LINUX [Linux 4.19.97-v7+ armv7l]
  Built: Tue Feb 25 2020 21:23:14 UTC
  Built With:
    configure  '--build=arm-linux-gnueabihf' '--includedir=${prefix}/include' '--mandir=${prefix}/share/man' '--infodir=${prefix}/share/info' '--sysconfdir=/etc' '--localstatedir=/var' '--disable-silent-rules' '--libdir=${prefix}/lib/arm-linux-gnueabihf' '--libexecdir=${prefix}/lib/arm-linux-gnueabihf' '--disable-maintainer-mode' '--disable-dependency-tracking' '--prefix=/usr' '--with-includes=/usr/include/postgresql:/usr/include/mariadb:/usr/include/mariadb/mysql' '--mandir=/usr/share/man' '--sysconfdir=/etc/proftpd' '--localstatedir=/run' '--libexecdir=/usr/lib/proftpd' '--enable-sendfile' '--enable-facl' '--enable-dso' '--enable-autoshadow' '--enable-ctrls' '--enable-ipv6' '--enable-nls' '--enable-memcache' '--with-lastlog=/var/log/lastlog' '--enable-pcre' '--disable-strip' '--enable-redis' '--build' 'arm-linux-gnueabihf' '--with-shared=mod_unique_id:mod_site_misc:mod_load:mod_ban:mod_quotatab:mod_sql:mod_sql_mysql:mod_sql_postgres:mod_sql_sqlite:mod_sql_odbc:mod_dynmasq:mod_quotatab_sql:mod_ldap:mod_quotatab_ldap:mod_ratio:mod_tls:mod_rewrite:mod_radius:mod_wrap:mod_wrap2:mod_wrap2_file:mod_wrap2_sql:mod_quotatab_file:mod_quotatab_radius:mod_facl:mod_ctrls_admin:mod_copy:mod_deflate:mod_ifversion:mod_geoip:mod_exec:mod_sftp:mod_sftp_pam:mod_sftp_sql:mod_shaper:mod_sql_passwd:mod_ifsession:mod_auth_otp:mod_tls_redis:mod_wrap2_redis:mod_redis:mod_memcache:mod_tls_memcache:mod_readme:mod_snmp' 'build_alias=arm-linux-gnueabihf' 'CFLAGS=-g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-Q9YVRO/proftpd-dfsg-1.3.6=. -fstack-protector-strong -Wformat -Werror=format-security' 'LDFLAGS=-Wl,-z,relro' 'CPPFLAGS=-Wdate-time -D_FORTIFY_SOURCE=2' 'CXXFLAGS=-g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-Q9YVRO/proftpd-dfsg-1.3.6=. -fstack-protector-strong -Wformat -Werror=format-security'

  CFLAGS: -g2 -g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-Q9YVRO/proftpd-dfsg-1.3.6=. -fstack-protector-strong -Wformat -Werror=format-security -Wall -fno-omit-frame-pointer -Werror=implicit-function-declaration
  LDFLAGS: -L$(top_srcdir)/lib -Wl,-z,relro -rdynamic  -L/usr/lib/arm-linux-gnueabihf/ -L/usr/lib/arm-linux-gnueabihf
  LIBS: -lacl  -lpcreposix -lpcre -lssl -lcrypto -lcap  -lpam -lsupp -lattr -lnsl -lresolv -lresolv -lcrypt -ldl -lhiredis -lmemcachedutil -lmemcached  -pthread

  Files:
    Configuration File:
      /etc/proftpd/proftpd.conf
    Pid File:
      /run/proftpd.pid
    Scoreboard File:
      /run/proftpd.scoreboard
    Header Directory:
      /usr/include/proftpd
    Shared Module Directory:
      /usr/lib/proftpd

  Info:
    + Max supported UID: 4294967295
    + Max supported GID: 4294967295

  Features:
    + Autoshadow support
    + Controls support
    + curses support
    - Developer support
    + DSO support
    + IPv6 support
    + Largefile support
    + Lastlog support
    + Memcache support
    + ncursesw support
    + NLS support
    + Redis support
    - Sodium support
    + OpenSSL support
    + PCRE support
    + POSIX ACL support
    + Shadow file support
    + Sendfile support
    + Trace support
    + xattr support

  Tunable Options:
    PR_TUNABLE_BUFFER_SIZE = 1024
    PR_TUNABLE_DEFAULT_RCVBUFSZ = 8192
    PR_TUNABLE_DEFAULT_SNDBUFSZ = 8192
    PR_TUNABLE_ENV_MAX = 2048
    PR_TUNABLE_GLOBBING_MAX_MATCHES = 100000
    PR_TUNABLE_GLOBBING_MAX_RECURSION = 8
    PR_TUNABLE_HASH_TABLE_SIZE = 40
    PR_TUNABLE_LOGIN_MAX = 256
    PR_TUNABLE_NEW_POOL_SIZE = 512
    PR_TUNABLE_PATH_MAX = 4096
    PR_TUNABLE_SCOREBOARD_BUFFER_SIZE = 80
    PR_TUNABLE_SCOREBOARD_SCRUB_TIMER = 30
    PR_TUNABLE_SELECT_TIMEOUT = 30
    PR_TUNABLE_TIMEOUTIDENT = 10
    PR_TUNABLE_TIMEOUTIDLE = 600
    PR_TUNABLE_TIMEOUTLINGER = 10
    PR_TUNABLE_TIMEOUTLOGIN = 300
    PR_TUNABLE_TIMEOUTNOXFER = 300
    PR_TUNABLE_TIMEOUTSTALLED = 3600
    PR_TUNABLE_XFER_SCOREBOARD_UPDATES = 10
```

**Comment by Castaglia** [comment]
It's better to use the entire ProFTPD source from the `master` branch, not just cherry-picking a single file, and assuming that that will work.  There are often things like Autoconf detection, other header changes involved with supporting different OpensSL library versions.

And if your current `proftpd` executable not linked against OpenSSL 1.1.1d, then trying to use `prxs` to compile a `mod_digest` DSO that _does_ link with OpenSSL 1.1.1d will probably not work that well.

**Comment by hpreusse** [comment]
@ardgeo: the issue has been solved in 1.3.7.
@Castaglia: the issue is reproducible in 1.3.6, but solved in 1.3.7.

**Comment by Castaglia** [comment]
Considering that this works with 1.3.7, I'm going to close this ticket.
